docs/optimization.md

```diff
@@ -0,0 +1,153 @@
+# A Developer's Guide to the Optimizer (SELECT version)
+
+Query optimization is the process by which a physical operator query plan is generated from an input SQL statement. In this document, we will go through the process of optimizing a simple `SELECT` statement in NoisePage. Other queries will follow a similar format, but might have different sets of rules explored.
+
+The example referred to in this guide is generated from the query
+
+```sql
+SELECT NO_O_ID FROM NEW_ORDER 
+WHERE NO_D_ID = 4 AND NO_W_ID = 1
+ORDER BY NO_O_ID ASC  
+LIMIT 5;
+```
+
+run on a loaded TPCC dataset with `ScaleFactor=4`.
+
+# 1. Traffic Cop Layer
+
+Optimization starts after the binding process in our system's traffic cop.
+
+In `traffic_cop.cpp`, `TrafficCop::OptimizeBoundQuery` takes as input a:
+
+- `const common::ManagedPointer<network::ConnectionContext> connection_ctx`: A context that stores the state of the active connection to the client
+- `const common::ManagedPointer<parser::ParseResult> query`: The output statements and expressions generated by the parser
+
+and will generate as output a:
+
+- `std::unique_ptr<planner::AbstractPlanNode> physical_plan`: the generated physical plan for the query
+
+This function, in turn, calls `TrafficCopUtil::Optimize` that takes in additional arguments to track the active transaction, catalog information, the internal database ID, statistics, the active cost model, and a timeout period for the optimizer, which explored further below.
+
+# 2. Input Transformation
+
+The first step within the optimization process is to transform the input `SQLStatement` to a set of logical expressions,`AbstractOptimizerNode logical_exprs`, in `QueryToOperatorTransformer::ConvertToOpExpression`.  In the `QueryToOperatorTransformer` class, we start by visiting the `SelectStatement` outputted by the parse process:
+
+- If the `SELECT` statement operates upon a table, the `TableRef` is accepted as well and added to the `output_expr_` as a `LogicalGet`. Otherwise, the `LogicalGet` is initialized to operate on an invalid table (used for queries like `SELECT 1 + 1`).
+- Predicates for the input expression, if they exist, are collected with an alias to each referenced column's associated table name in `std::vector<AnnotatedExpression> predicates_` and are stored in a `LogicalFilter` that is added as the parent of the existing `output_expr_`
+    - [ ]  Add information for subquery transformation above
+- [ ]  Add information about aggregations in `SELECT` here
+- [ ]  Add information about `SelectDistinct` in `SELECT` here
+- If a `LIMIT` exists in the select statement, the value of the limit and associated offset, along with any  `ORDER_BY` clauses, are stored in a `LogicalLimit` node that is added as the parent of the existing `output_expr_`
```

Comment:
I think it would be helpful here to briefly explain why `ORDER_BY` information is stored within a `LIMIT` operator